crates/red_knot/src/types/infer.rs

```diff
@@ -145,6 +145,15 @@ fn infer_expr_type(db: &dyn SemanticDb, file_id: FileId, expr: &ast::Expr) -> Qu
     // TODO cache the resolution of the type on the node
     let symbols = symbol_table(db, file_id)?;
     match expr {
+        ast::Expr::NumberLiteral(ast::ExprNumberLiteral { value: v, .. }) => {
+            if let ast::Number::Int(n) = v {
+                // TODO support big int literals, or at least default to `builtins.int`
+                Ok(n.as_i64().map(Type::IntLiteral).unwrap_or(Type::Unknown))
+            } else {
+                // TODO builtins.float or builtins.complex
```

Comment:
`builtins.complex` and `builtins.float` are not valid parameters to `Literal` in type expressions. But I suppose for our purposes, we need to hang on to the actual value of the symbol as well as the type, as many lint rules may find the actual value useful. But we may need to think carefully about how we display these types to users in diagnostic messages -- it might be confusing if we reported to a user that a variable has type `Literal[3.1234552]` or whatever, when that isn't a valid Python type according to the specification.

I think mypy does something similar where they have an internal representation for `Literal` floats; this is important for when they want to compile Python code with mypyc (not a concern for us). But that Literal type is never shown to users.

Reply:
`builtins.float` and `builtins.complex` here are meant to simply say that we should infer those general types (the type `float` or the type `complex`), not a `Literal` of those types.

I agree we may want to consider internally supporting literals of them as well, but this comment wasn't meant to imply anything about that.

Is there a way to word it that would make that clearer?

Reply:
Right, sorry -- I was reading the TODO comment in the context of the PR's purpose and assumed you meant something to do with `Literal` types. But in the context of the function I think it's clear as-is; I just didn't look closely enough.